### PR TITLE
add TaskDefinition Info for global service task handler

### PR DESCRIPTION
**Pull request by averyyan**
i am try to use TaskDefinition info for automatic match grpc job service.
I want to replace AddTaskHandler to job service. I think id map is not enough.

### CODE REVIEW

**Comment by nitram509** [comment]
Hi @averyyan 
thank you for this PR.
I got your point, that adding Task-Handlers by ID is not ideal, and you're seeking for more global, or flexible way of defining Task Hnadlers.
I don't like the idea of a fixed serviceTaskHandler, as this bring some degree of flexibility, but usually, global definitions also have their down-sides.

What about "task type", see https://docs.camunda.io/docs/components/modeler/bpmn/service-tasks/
Also, please check the Camunda Modeller (model a v8 bpmn) and see how task type work.
This is something I would prefer to implement instead, since it would be more compatible with Zeebe and also gives more flexible way of defining a handler to one or multiple tasks - especially meaningful, when we later support sub-processes etc.

Would that task type work for your use case as well?

PS: when you consider refactoring your PR towards task type support - that would be awesome -
, please let us keep the existing approach.
I would prefer co-existence of ```addTaskHandlerForId()``` and ```addTaskHandlerForType()```.

**Comment by averyyan** [comment]
> Hi @averyyan thank you for this PR. I got your point, that adding Task-Handlers by ID is not ideal, and you're seeking for more global, or flexible way of defining Task Hnadlers. I don't like the idea of a fixed serviceTaskHandler, as this bring some degree of flexibility, but usually, global definitions also have their down-sides.
> 
> What about "task type", see https://docs.camunda.io/docs/components/modeler/bpmn/service-tasks/ Also, please check the Camunda Modeller (model a v8 bpmn) and see how task type work. This is something I would prefer to implement instead, since it would be more compatible with Zeebe and also gives more flexible way of defining a handler to one or multiple tasks - especially meaningful, when we later support sub-processes etc.
> 
> Would that task type work for your use case as well?
> 
> PS: when you consider refactoring your PR towards task type support - that would be awesome - , please let us keep the existing approach. I would prefer co-existence of `addTaskHandlerForId()` and `addTaskHandlerForType()`.

I just have this idea from  https://docs.camunda.io/docs/components/modeler/bpmn/service-tasks/. 
I think task definition is for definite kind of jobs.
I change this to addTaskHandlerForType() , it works too.
I use global fun because I have a problem to decide `addTaskHandlerForId()` ,`addTaskHandlerForType()` which one should be first :)

File: pkg/bpmn_engine/engine.go [hunk 1]
```diff
@@ -286,6 +287,11 @@ func checkOnlyOneAssociationOrPanic(event *BPMN20.BaseElement) {
 	}
 }
 
+// SetServiceTaskHandler set global service task handler to use task definition info
+func (state *BpmnEngineState) SetServiceTaskHandler(handler func(job ActivatedJob)) {
```
Comment (nitram509) [inline pkg/bpmn_engine/engine.go:291]: This func is not mentioned in the interface.

File: pkg/spec/BPMN20/model.go [hunk 2]
```diff
@@ -81,6 +82,11 @@ type TUserTask struct {
 	Output              []TIoMapping `xml:"extensionElements>ioMapping>output"`
 }
 
+type TTaskDefinition struct {
+	TypeName string `xml:"type,attr"`
+	Retries  string `xml:"retries,attr"`
```
Comment (nitram509) [inline pkg/spec/BPMN20/model.go:87]: Since 'retries' is not used yet, I would prefer to not include it yet.

File: pkg/bpmn_engine/engine_test.go [hunk 3]
```diff
@@ -166,3 +166,40 @@ func TestParallelGateWayTwoTasks(t *testing.T) {
...
+	process, _ := bpmnEngine.LoadFromFile("../../test-cases/simple_task.bpmn")
+	// when
+	bpmnEngine.CreateAndRunInstance(process.ProcessKey, nil)
+
+	definition := bpmnEngine.processes[0].definitions.Process.ServiceTasks[0].TaskDefinition
+	then.AssertThat(t, definition.TypeName, is.EqualTo("TestType"))
+	then.AssertThat(t, definition.Retries, is.EqualTo("testValue"))
+}
+
+func TestServiceTaskHandler(t *testing.T) {
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_test.go:182]: that test is not easy to understand and lacks expressiveness, how service tasks are supposed to work.
I do suggest creating a dedicated .bpmn file to showcase service tasks

**Review by nitram509 — COMMENTED** [review]
Hi,
after carefully rethinking the overall approach of a single service handler, I would like to reject this PR. Multiple thoughts made me conclude this.
First, I was inspired by your idea. Also, I fully anticipate the use case you're describing.
In general, I prefer some degree of compatibility with Zeebe. While having a single global handler, which might fit your use case, others might want to use e.g. 3 handlers for 10 tasks. This is/was the reason, why Zeebe uses taskDefintion.Type to give flexibility in the assignments.
So, I thought a lot and had so many ideas for an alternative implementation, that I gave it a try on my own. For formal reasons, I also created a new issue #58 .

I would be glad if you could have a look at the implementation and check if it fits your use case, please. (I plan to push my code changes today)

**Comment by nitram509** [comment]
Not merged, in favor of https://github.com/nitram509/lib-bpmn-engine/commit/5303ef37f3f3c141d62dadd4fff227c3ed9b1f63

**Pull request closed without merging**